### DISCUSSION TITLE
FastAPI - restApi is not database async?

### DISCUSSION THREAD

**Discussion by vlori2k** [fastapi/fastapi#6579] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
#crud.py:

from sqlalchemy.orm import Session

from fastapi_restapi import models, schemas
from sqlalchemy.sql import func



def get_all_students_from_db(db: Session):
    return db.query(models.Students).all()

#-------------------------------------------------
#database.py:

import os

from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker



the_db_address = os.getenv("DB_ADDRESS").replace("'", "")
the_db_port = os.getenv("DB_PORT").replace("'", "")
the_db_user = os.getenv("DB_USER").replace("'", "")
the_db_password = os.getenv("DB_PASSWORD").replace("'", "")
the_database_to_use = os.getenv("WOIT_DATA_DB").replace("'", "")

complete_sql_string = the_db_user + ":" + the_db_password + "@" + the_db_address + "/" +the_database_to_use

SQLALCHEMY_DATABASE_URL = "mysql+pymysql://" + complete_sql_string

engine = create_engine(SQLALCHEMY_DATABASE_URL)

SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

SessionLocal_async = sessionmaker(autocommit=True, autoflush=True, bind=engine)

Base = declarative_base()



def normal_DB():
    try:
        db = SessionLocal()
        yield db
    finally:
        db.close()


def async_DB(): #used for always having new data available
    try:
        db = SessionLocal_async()
        yield db
    finally:
        db.close()
#-------------------------------------------------
#main.py:
from fastapi.middleware.cors import CORSMiddleware
from sqlalchemy.orm import Session

from fastapi_restapi import models, schemas, crud


from .database import engine, normal_DB, async_DB
import asyncio
from fastapi import FastAPI, HTTPException, Security
from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer



models.Base.metadata.create_all(bind=engine)


async def get_all_the_students(db: Session = next(async_DB())):
    while True:
        await asyncio.sleep(10)
        all_studs = crud.get_all_students_from_db(db)
        for i in all_studs:
            print(i.student_name)
            #tim
            #rick
            #anna



students_from_db = asyncio.ensure_future(get_all_the_students())
```

### Description

I thought this problem was solved, but its not..

Because of large amount of data inside the db, im planning to store some of it on the server.

The function inside main.py "pokes" the the DB each 10 seconds, and gets the data, which works fine, and i can see it, like this:
            tim
            rick
            anna

But if some changes are made inside the DB, the server keeps poking the DB again, but i can not see the changes? Lets say i change the name "tim" to "jonny", then then the server pokes the DB again after 10 sec i still see this:
            tim
            rick
            anna
            
Is there anyone here who can help solving this issue?

### Operating System

Linux

### Operating System Details

Linux ubuntu , and the restAPI is up and running using docker

### FastAPI Version

0.68.0

### Python Version

3.8

**Comment by dstlny** [comment]
This isn't a FastAPI issue.
first of all:
`db: Session = next(async_DB())`
"async_DB" is sync, and so you are only using the one session from when the get_all_the_students coroutine was initially running. It hasn't closed, thus it hasn't committed - or flushed the connection, and you're still getting the old data.

**Comment by provinzkraut** [comment]
You're not setting up an async `Session`. Simply naming it that won't magically make it async ;)
You need to use the actual [`AsyncSession`](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncSession) class and [`AsyncEngine`](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncEngine) as well. See the SQLAlchemy documentation for details on how to set this all up: https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html

You also should define `db: Session` as a dependency, as [described in the docs](https://fastapi.tiangolo.com/tutorial/sql-databases/#create-a-dependency).

**Comment by vlori2k** [comment]
> This isn't a FastAPI issue.
> first of all:
> `db: Session = next(async_DB())`
> "async_DB" is sync, and so you are only using the one session from when the get_all_the_students coroutine was initially running. It hasn't closed, thus it hasn't committed - or flushed the connection, and you're still getting the old data.

My apologies then, I have seen some example were people use db: Sesssion in their fastApi codes, so i thought i was a part of this :)

**Comment by vlori2k** [comment]
> You're not setting up an async `Session`. Simply naming it that won't magically make it async ;)
> You need to use the actual [`AsyncSession`](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncSession) class and [`AsyncEngine`](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncEngine) as well. See the SQLAlchemy documentation for details on how to set this all up: https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html
> 
> You also should define `db: Session` as a dependency, as [described in the docs](https://fastapi.tiangolo.com/tutorial/sql-databases/#create-a-dependency).

Thank you so much for information. I may close this issue then :)
